core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java

```diff
@@ -179,37 +190,54 @@ public void write(Iterator<Product2<K, V>> records) throws IOException {
    *
    * @return array of lengths, in bytes, of each partition of the file (used by map output tracker).
    */
-  private long[] writePartitionedFile(File outputFile) throws IOException {
+  private long[] writePartitionedData(ShuffleMapOutputWriter mapOutputWriter) throws IOException {
     // Track location of the partition starts in the output file
     final long[] lengths = new long[numPartitions];
     if (partitionWriters == null) {
       // We were passed an empty iterator
       return lengths;
     }
-
-    final FileOutputStream out = new FileOutputStream(outputFile, true);
     final long writeStartTime = System.nanoTime();
-    boolean threwException = true;
     try {
       for (int i = 0; i < numPartitions; i++) {
         final File file = partitionWriterSegments[i].file();
-        if (file.exists()) {
-          final FileInputStream in = new FileInputStream(file);
-          boolean copyThrewException = true;
-          try {
-            lengths[i] = Utils.copyStream(in, out, false, transferToEnabled);
-            copyThrewException = false;
-          } finally {
-            Closeables.close(in, copyThrewException);
+        boolean copyThrewException = true;
+        ShufflePartitionWriter writer = null;
+        try {
```

Comment:
nit: i think you can reduce the amount of repeated code and nesting by moving the check for `file.exists()` to right after line 203
```
final File file = partitionWriterSegments[i].file();
if (!file.exists()) continue
boolean copyThrewException = true
ShufflePartitionWriter writer = null
try {
  writer = mapOutputWriter.getNextPartitionWriter();
  try {
    FileInputStream in = new FileInputStream(file)
    if (transferToEnabled) {
      WritableByteChannel outputChannel = writer.toChannel();
      FileChannel inputChannel = in.getChannel()
      Utils.copyFileStreamNIO(inputChannel, outputChannel, 0, inputChannel.size());
      copyThrewException = false;
    } else {
      OutputStream tempOutputStream = writer.toStream();
      Utils.copyStream(in, tempOutputStream, false, false);
      copyThrewException = false;
    }
    
  } finally {
    Closeables.close(in, copyThrewException);
  }
  if (!file.delete()) {
    logger.error();
  }
} finally {
  Closeables.close(writer, copyThrewException);
}